### PR TITLE
Add dispersion parameter projection method

### PR DESCRIPTION
**Pull request by yannmclatchie**
## Description

This PR introduces the following updates:
1. Allow for dispersion parameters to be projected when included in the reference model family
2. Change the way a ``Family`` object is instantiated (using a bambi model instead of a string)
3. Reorganise the project to have the ``Projector`` object exposed from a ``base`` file
4. Add more attributes to the ``ModelData`` object

## Next steps

We now look towards building the ``search`` method which is an integral part of the model selection procedure. Before we proceed, the following features should be developed.
- Add R-hat convergence checks for projected parameters
- More robust posterior building for the restricted model
- Compute and add ``ELPD`` and distance to reference model to ``res_model`` object
- Redefine ``cov_names`` attribute of ``ModelData`` object to be more consistent (include ``Intercept`` term if necessary for example) and add ``disp_name`` for clarity

### CODE REVIEW

File: src/kulprit/base.py [hunk 1]
```diff
@@ -1,33 +1,37 @@
...
-    def __init__(self, model, posterior):
+    def __init__(self, model, posterior=None):
         """Reference model builder for projection predictive model selection.
 
         This object initialises the reference model and handles the core
         projection and variable search methods of the model selection procedure.
 
         Args:
-            model (bambi.models.Model): The Bambi GLM model of interest
+            model (bambi.models.Model): The referemce GLM model to project
```
Comment (aloctavodia) [inline src/kulprit/base.py:25]:
```suggestion
            model (bambi.models.Model): The reference GLM model to project
```

File: src/kulprit/base.py [hunk 2]
```diff
@@ -1,33 +1,37 @@
...
     _extract_insample_predictions,
+    _build_posterior,
+    _compute_elpd,
 )
-from ..families import Family
 
 
 class Projector:
-    def __init__(self, model, posterior):
+    def __init__(self, model, posterior=None):
```
Comment (aloctavodia) [inline src/kulprit/base.py:18]: instead of `posterior` use  `idata` or `inferencedata`

File: src/kulprit/base.py [hunk 3]
```diff
@@ -1,33 +1,37 @@
...
         """Reference model builder for projection predictive model selection.
 
         This object initialises the reference model and handles the core
         projection and variable search methods of the model selection procedure.
 
         Args:
-            model (bambi.models.Model): The Bambi GLM model of interest
+            model (bambi.models.Model): The referemce GLM model to project
             posterior (arviz.InferenceData): The posterior arViz object of the
```
Comment (aloctavodia) [inline src/kulprit/base.py:26]:
```suggestion
            posterior (arviz.InferenceData): The posterior ArviZ object of the
```

File: src/kulprit/base.py [hunk 4]
```diff
@@ -37,21 +41,32 @@ def __init__(self, model, posterior):
...
             y=y,
             data=data,
             link=link,
             family=family,
             cov_names=cov_names,
+            response_name=response_name,
             n=n,
             m=m,
             s=s,
             has_intercept=has_intercept,
```
Comment (aloctavodia) [inline src/kulprit/base.py:68]: what about using variable names more descriptive than n, m, s?

File: src/kulprit/families/family.py [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+
+    @classmethod
+    def create(cls, model):
+        if model.family.name not in cls.subclasses:
+            raise NotImplementedError("Unsupported family.")
+
+        return cls.subclasses[model.family.name]()
+
+
+class Gaussian(Family):
```
Comment (aloctavodia) [inline src/kulprit/families/family.py:33]: Maybe not important to change right away, but... If we are using something general as the Wasserstein distance, we may not need to have families. And if we want to provide more than one distance/loss option (for us or users) in the future then we could already start organizing the code around distances/losses.

File: src/kulprit/utils.py [hunk 6]
```diff
@@ -48,23 +48,43 @@ def _extract_insample_predictions(model):
     return y_pred
 
 
-def _extract_theta_perp(solver, cov_names):
-    """Extract restricted parameter projections from PyTorch optimisation.
+def _build_posterior(theta_perp, model, disp_perp=None):
+    """Convert some set of pytorch tensors into an arViz InferenceData object.
```
Comment (aloctavodia) [inline src/kulprit/utils.py:52]:
```suggestion
    """Convert some set of pytorch tensors into an ArviZ InferenceData object.
```

File: src/kulprit/base.py (hunk 4 above)
Comment (yannmclatchie) [reply src/kulprit/base.py:68]: Yes, I think this would help the legibility of code. How about ``num_obs``, ``num_params``, ``num_draws``?

File: src/kulprit/families/family.py (hunk 5 above)
Comment (yannmclatchie) [reply src/kulprit/families/family.py:33]: I think splitting it by family will allow us the most flexibility in dealing with which distance is best for each model type, since we may not use Wasserstein distance for all distributions (e.g. I think we can still use KL for binomial). Something similar in ``projpred`` in ``extend_family.R``. Having said that, if we can find a general distance formulation which can acts as a surrogate for KL divergence in order to remain close to the procedure as initially outlined in Goutis and Robert (1998), then perhaps having a unified loss class would be best.

File: src/kulprit/base.py (hunk 2 above)
Comment (yannmclatchie) [reply src/kulprit/base.py:18]: Yes using something like ``inferencedata`` would be clearer, especially since later the ``_build_posterior()`` method and others may do more than simply extract posterior draws sto allow for more parsimonious ``ModelData`` classes.

File: README.md [hunk 7]
```diff
@@ -40,16 +38,14 @@ data = pd.DataFrame({
 })
 # define and fit model with MCMC
 model = bmb.Model("y ~ x1 + x2", data, family="gaussian")
-num_draws, num_chains = 100, 1
-posterior = model.fit(draws=num_draws, chains=num_chains)
+posterior = model.fit()
```
Comment (aloctavodia) [inline README.md:41]:
```suggestion
idata = model.fit()
```

File: README.md [hunk 8]
```diff
@@ -40,16 +38,14 @@ data = pd.DataFrame({
...
 # build reference model object
-ref_model = kpt.Projector(model, posterior)
-# project the reference model to some parameter subset
-params = ["x1", "x2"]
-theta_perp = ref_model.project(params=params)
-print(theta_perp.shape)
-# visualise the projected model posterior
-ref_model.plot_projection(params=params)
+proj = kpt.Projector(model, posterior)
```
Comment (aloctavodia) [inline README.md:43]:
```suggestion
proj = kpt.Projector(model, idata)
```

File: src/kulprit/base.py [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+            num_draws=num_draws,
+            has_intercept=has_intercept,
+            dist_to_ref_model=dist_to_ref_model,
+            inferencedata=inferencedata,
+            predictions=predictions,
+        )
+
+    def project(
+        self,
+        cov_names=None,
```
Comment (aloctavodia) [inline src/kulprit/base.py:78]: To make the argument's name more consistent with ArviZ (and hence pymc/bambi) we may want to use `var_names`

**Review by aloctavodia — APPROVED** [review]

**Review by aloctavodia — APPROVED** [review]

File: src/kulprit/base.py (hunk 9 above)
Comment (yannmclatchie) [reply src/kulprit/base.py:78]: This sounds like a good idea, I'll make the changes

Comment (yannmclatchie) [reply src/kulprit/base.py:78]: I am currently working on a new definition of ``var_names``/``cov_names`` such that it includes the ``Intercept`` term, and then also changing the definition of the ``project`` method to more closely resemble it's [R counterpart](https://mc-stan.org/projpred/reference/project.html), in other words a user will call
```python
proj.project(num_vars=2) # or perhaps proj.project(model_size=2)
```
instead of
```python
proj.project(var_names=["x1", "x2"])
```
For now (until a search method is implemented), this will probably just select the first ``num_vars`` of the reference model, and the next step will be to compute the ELPD of different models as a first step towards a search method. 

Since these two things (renaming ``cov_names`` and redefining ``project``) are related, perhaps I will merge this branch and open a new PR with those features?

Comment (aloctavodia) [reply src/kulprit/base.py:78]: sure

Comment (yannmclatchie) [reply src/kulprit/base.py]: So in fact I've added those features to this branch as they didn't take much more work. 

The one thing to consider now is when a user calls
```python
proj.project(model_size=1)
```
should this be a submodel with an intercept and one variable, or just an intercept? As I understand from @fweber144, in ``r-projpred``, if a user calls
```R
projpred::project(ref_model, nterms = 1)
```
this will return a model with an intercept and one variable. In the new documentation there a note reading

>Note that ``nterms`` does not count the intercept, so use ``nterms = 0`` for the intercept-only model.

It seems that this design choice is to allow the package to deal with reference models that do not have intercepts equivalently to those that do. It may then be a good idea to change the current ``model_size`` parameter to ``num_vars`` and use the same definition as in ``projpred``. Let me know what you think.

Comment (aloctavodia) [reply src/kulprit/base.py]: This makes sense in general and in particular it also align with Bambi.

**Comment by aloctavodia** [comment]
I understand this project is in early stages so everything is subject to massive changes. But I think is a good idea, if possible, to have PR with small changes that are merge reasonably fast.
